BWMatch.py

```diff
@@ -0,0 +1,139 @@
+'''
+Implement efficient algos from Pevzner & Compeau BWT chapter
+Construct BWT from suffix tree
+Suffix tree:all nodes are labeled with the start and end locations in T corresponding to their text string
+Leaf nodes are additionally labeled with the index of where the suffix ending there originated in T
+Then to find the BWT, we do a DFS where we always start by exploring the child node with the earlier character in the alphabet. And when we get to a leaf, we append T[i - 1] to the growing BWT.
+
+Data structures:
+index everything by integers: separate dictionary for each piece of data
+G gives the child nodes, NS, NE give the start and end locations of the node substrings, SS gives the suffix starting indices
+
+'''
+
+from graphviz import Digraph
+import time
+import re
+import argparse
+
+class BWMatch:
+	def __init__(self):
+		pass
+
+	def loadIndex(self, filename):
+		f = open(filename)
+		self.BWT = f.readline().strip()
+		SA = f.readline().strip().split(',')
+		self.suffixArray = {}
+		for s in SA:
+			k, v = s.split(';')
+			self.suffixArray[int(k)] = int(v)
+		alphabet = [x for x in f.readline().strip().split(',')]
+		self.C = {}
+		for a in alphabet:
+			self.C[a] = [int(x) for x in f.readline().strip().split(',')]
+		self.FO = {}
+		FO_items = [int(x) for x in f.readline().strip().split(',')]
+		for i, a in enumerate(alphabet):
+			self.FO[a] = FO_items[i]
+		self.SA_gap = int(f.readline().strip())
+		self.C_gap = int(f.readline().strip())
+
+	def lastToFirst(self, index):
+		letter = self.BWT[index]
+		return self.FO[letter] + self.reconstructCount(letter, index)
+
+	def reconstructSuffixArray(self, index):
+		count = 0
+		while self.suffixArray.get(index) is None:
+			index = self.lastToFirst(index)
+			count += 1
+		return self.suffixArray[index] + count
+
+	def FindMatches(self, Pattern):
+		top = 0
+		bottom = len(self.BWT) - 1
+		while top <= bottom:
+			if len(Pattern) > 0:
+				symbol = Pattern[-1]
+				Pattern = Pattern[:-1]
+				C_top = self.reconstructCount(symbol, top)
+				C_bottom = self.reconstructCount(symbol, bottom + 1)
+				if C_bottom - C_top > 0:
+					top = self.FO[symbol] + C_top
+					bottom = self.FO[symbol] + C_bottom - 1
+				else:
+					return []
+			else:
+				return [self.reconstructSuffixArray(x) for x in range(top, bottom + 1)]
+
+	def MatchAllPatterns(self, patternfile, outputfile):
+		g = open(outputfile, "w")
+		with open(patternfile) as f:

```

Comment:
Here you can (and probably should, so that the outputfile is closed) write both of these within the `with`:
```python
with open(patternfile) as f, open(outputfile, "w") as g:
```